File: Objective C/Direct Integration/OptInVideoExample/OptInVideoExample/ViewController.m
//
//  ViewController.m
//  OptInVideoExample
//
//  Copyright © 2020 Ogury Co. All rights reserved.
//

#import "ViewController.h"
#import <OguryChoiceManager/OguryChoiceManager.h>
#import <OguryAds/OguryAds.h>

@interface ViewController()<OguryAdsOptinVideoDelegate>

@property (nonatomic, weak) IBOutlet UITextView *statusTextView;
@property (nonatomic, strong) OguryAdsOptinVideo *optinVideo;
@property (nonatomic, assign) BOOL isAdLoaded;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self addNewStatus: @"Choice Manager Loading..."];
    
    //The setup of Ogury Choice Manager and Ogury Ads is done AppDelegate file.
    [[OguryChoiceManager sharedManager] askWithViewController:self andCompletionBlock:^(NSError * _Nullable error, OguryChoiceManagerAnswer answer) {
        if (!error) {
            switch (answer) {
                case OguryChoiceManagerAnswerNoAnswer:
                    [self addNewStatus: @"Choice Manager No Answer"];
                    break;
                case OguryChoiceManagerAnswerFullApproval: // TCF Option
                    [self addNewStatus: @"Choice Manager Full Approval"];
                    break;
                case OguryChoiceManagerAnswerPartialApproval: // TCF Option
                    [self addNewStatus: @"Choice Manager Partial Approval"];
                    break;
                case OguryChoiceManagerAnswerRefusal: // TCF Option
                    [self addNewStatus: @"Choice Manager Refusal"];
                    break;
                case OguryChoiceManagerAnswerSaleAllowed: // CCPA Option
                    [self addNewStatus: @"Choice Manager Sale Allowed"];
                    break;
                case OguryChoiceManagerAnswerSaleDenied: // CCPA Option
                    [self addNewStatus: @"Choice Manager Unknown Option"];
                    break;
            }
        } else {
            [self addNewStatus: [NSString stringWithFormat:@"Choice Manager error : %@", error.description]];
        }
    }];
    // Do any additional setup after loading the view.
}

- (IBAction)loadAdBtnPressed:(id)sender {
    [self addNewStatus: @"Loading Ad..."];
    self.optinVideo = [[OguryAdsOptinVideo alloc]initWithAdUnitID:@"ogury_adunit"];
    self.optinVideo.optInVideoDelegate = self;
    [self.optinVideo load];
}

- (IBAction)showAdBtnPressed:(id)sender {
    if (self.isAdLoaded == YES) {
        [self addNewStatus: @"Ad requested to show"];
        [self.optinVideo showAdInViewController:self];
    } else {
        [self addNewStatus: @"Ad not loaded"];
    }
}


- (void)addNewStatus:(NSString *)status {
    NSString * statusLog = [status stringByAppendingString:@"\n"];
    self.statusTextView.text = [self.statusTextView.text stringByAppendingString:statusLog];
    NSRange bottom = NSMakeRange(self.statusTextView.text.length-1, 1);
    [self.statusTextView scrollRangeToVisible:bottom];
}

#pragma mark - OguryAds Delegate

- (void)oguryAdsOptinVideoAdLoaded {
    [self addNewStatus: @"Ad received"];
    self.isAdLoaded = YES;
}

- (void)oguryAdsOptinVideoAdClosed {
    [self addNewStatus: @"Ad Closed"];
    self.isAdLoaded = NO;
}

- (void)oguryAdsOptinVideoAdDisplayed {
    [self addNewStatus: @"Ad on screen"];
}

- (void)oguryAdsOptinVideoAdClicked {
    [self addNewStatus: @"Ad Clicked"];
}

- (void)oguryAdsOptinVideoAdNotAvailable {
    [self addNewStatus: @"Ad Not Available"];
}

- (void)oguryAdsOptinVideoAdError:(OguryAdsErrorType)errorType {
    [self addNewStatus: [NSString stringWithFormat:@"Error : %ld", errorType]];
    self.isAdLoaded = NO;
}

- (void)oguryAdsOptinVideoAdRewarded:(OGARewardItem *)item {
    [self addNewStatus: [NSString stringWithFormat: @"Reward received with name: : %@ and value : %@", item.rewardName,item.rewardValue]];
}


@end

File: Swift/Direct Integration/OptInVideoExample/OptInVideoExample/ViewController.swift
//
//  ViewController.swift
//  OptinVideoExample
//
//  Copyright © 2020 Ogury Co. All rights reserved.
//

import UIKit
import OguryChoiceManager
import OguryAds

class ViewController: UIViewController {
    
    @IBOutlet weak var statusTextView: UITextView!
    var rewardedAd: OguryAdsOptinVideo?
    var adLoaded: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.addNewStatus("Choice Manager Loading...")
        
        //The setup of Ogury Choice Manager and Ogury Ads is done AppDelegate.swift file.

        OguryChoiceManager.shared().ask(with: self) { (error, answer) in
            if error == nil {
                switch answer {
                case .noAnswer: // TCF Option
                    self.addNewStatus("Choice Manager No Answer")
                case .fullApproval: // TCF Option
                    self.addNewStatus("Choice Manager Full Approval")
                case .partialApproval: // TCF Option
                    self.addNewStatus("Choice Manager Partial Approval")
                case .refusal: // TCF Option
                    self.addNewStatus("Choice Manager Refusal")
                case .saleAllowed: // CCPA Option
                    self.addNewStatus("Choice Manager Sale Allowed")
                case .saleDenied: // CCPA Option
                    self.addNewStatus("Choice Manager Sale Denided")
                default:
                    self.addNewStatus("Choice Manager Unknown Option")
                }
            } else {
                self.addNewStatus("Choice Manager error : \(error.debugDescription)")
            }
        }
    }
    
    @IBAction func loadAdBtnPressed(_ sender: Any) {
        self.addNewStatus("Loading Ad...")
        
        rewardedAd = OguryAdsOptinVideo.init(adUnitID: "ogury_adunit")
        guard let rewardedAd = rewardedAd else {
            self.addNewStatus("Error while initialising the ad")
            return
        }
        rewardedAd.optInVideoDelegate = self
        rewardedAd.load()
    }
    
    @IBAction func showAdBtnPressed(_ sender: Any) {
        guard let rewardedAd = rewardedAd, adLoaded else {
            self.addNewStatus("Ad not loaded")
            return
        }
        self.addNewStatus("Ad requested to show")
        rewardedAd.showAd(in: self)
    }

    func addNewStatus(_ status: String) {
        DispatchQueue.main.async {
            let textToLog = status + "\n"
            self.statusTextView.textStorage.append(NSAttributedString(string: textToLog))
            let bottom = NSMakeRange(self.statusTextView.text.count - 1, 1)
            self.statusTextView.scrollRangeToVisible(bottom)
        }
    }

}

extension ViewController:OguryAdsOptinVideoDelegate {
    
    func oguryAdsOptinVideoAdLoaded() {
        self.addNewStatus("Ad received")
        self.adLoaded = true
    }
    
    func oguryAdsOptinVideoAdClosed() {
        self.addNewStatus("Ad Closed")
        self.adLoaded = false
    }
    
    func oguryAdsOptinVideoAdClicked() {
        self.addNewStatus("Ad Clicked")
    }
    
    func oguryAdsOptinVideoAdDisplayed() {
        self.addNewStatus("Ad on screen")
    }
    
    func oguryAdsOptinVideoAdNotAvailable() {
        self.addNewStatus("Ad Not Available")
    }
    
    func oguryAdsOptinVideoAdRewarded(_ item: OGARewardItem!) {
        print("Reward received with name: \(item.rewardName) and value: \(item.rewardValue) ")
    }
    
    //For more informations about error codes please refer to our documentation at https://docs.ogury.co/
    func oguryAdsOptinVideoAdError(_ errorType: OguryAdsErrorType) {
        
    }
}

